File: dotnet-desktop-guide/winforms/migration/snippets/clipboard-dataobject-net10/net/csharp/ITypedDataObjectExamples.cs
using System;
using System.Reflection.Metadata;
using System.Windows.Forms;

namespace ClipboardExamples
{
    public class DragDropExamples
    {
        public class MyItem
        {
            public string Name { get; set; }
            public string Value { get; set; }
        }

        // <DragDropUsage>
        private void OnDragDrop(object sender, DragEventArgs e)
        {
            if (e.Data is ITypedDataObject typedData)
            {
                // Retrieve files from drag data using a standard format
                if (typedData.TryGetData(DataFormats.FileDrop, out string[] files))
                    Console.WriteLine($"Dropped files: {string.Join(", ", files)}");

                // Retrieve text using a standard format
                if (typedData.TryGetData(DataFormats.Text, out string text))
                    Console.WriteLine($"Dropped text: {text}");

                // Retrieve custom items using an application-specific format
                if (typedData.TryGetData("CustomItem", out MyItem item))
                    Console.WriteLine($"Dropped custom item: {item.Name} = {item.Value}");
            }
        }
        // </DragDropUsage>
    }
}

File: dotnet-desktop-guide/winforms/migration/snippets/clipboard-dataobject-net10/net/vb/ITypedDataObjectExamples.vb
Imports System
Imports System.Reflection.Metadata
Imports System.Windows.Forms

Namespace ClipboardExamples

    Public Class DragDropExamples
        Public Class MyItem
            Public Property Name As String
            Public Property Value As String
        End Class

        ' <DragDropUsage>
        Private Sub OnDragDrop(sender As Object, e As DragEventArgs)
            If TypeOf e.Data Is ITypedDataObject Then
                Dim typedData As ITypedDataObject = CType(e.Data, ITypedDataObject)

                ' Retrieve files from drag data using a standard format
                Dim files As String() = Nothing
                If typedData.TryGetData(DataFormats.FileDrop, files) Then
                    Console.WriteLine($"Dropped files: {String.Join(", ", files)}")
                End If

                ' Retrieve text using a standard format
                Dim text As String = Nothing
                If typedData.TryGetData(DataFormats.Text, text) Then
                    Console.WriteLine($"Dropped text: {text}")
                End If

                ' Retrieve custom items using an application-specific format
                Dim item As MyItem = Nothing
                If typedData.TryGetData("CustomItem", item) Then
                    Console.WriteLine($"Dropped custom item: {item.Name} = {item.Value}")
                End If
            End If
        End Sub
        ' </DragDropUsage>
    End Class
End Namespace
